### ISSUE TITLE
New Collaborative Filtering Plugin

### ISSUE THREAD

**Issue by jackransomlovell** [jspsych/jsPsych#1881]
Hi all, (especially @becky-gilbert @jodeleeuw)

I have been trying my hand at developing a collaborative filtering plugin but am currently stuck. I have forked the repo and committed two files called ```jspych-collab-filtering*``` to the parent directory. One is the plugin (the .js file) and another is my attempt at a demo. The demo fails with an inability to find the plugin from ```jspsych.js``` which I could not figure out. So I guess first things first is figuring out why that is so. I based my thinking on the video-keyboard* and multi-slider* plugins. 

Additionally, I am wondering if my code in the .js file is redundant and how to maybe make it cleaner and modular. As the trial should have 4 components: the first viewing of the video stopped by a random pause, a multi-slider rating section, the resume of the video at the random time point, and then at the end of the video another multi-slider rating portion. Currently, I define HTML elements for each then just call ```display_element.innerHTML = ''``` at the end of each of the 4 sections to display the next part of the trial. The main contribution I have made is the ```random_stop()``` function, which defines the random stopping point for the video. 

Any help on the above would be very much appreciated, as I am certainly a novice when it comes to this line of programming :).

**Comment by becky-gilbert** [comment]
Hi @jackransomlovell! Is this a plugin that you'd like to contribute to the library? (Just asking so that I know whether or not to categorize this issue as a potential new plugin).

I had a look at your [collab-filter-dev branch](https://github.com/jackransomlovell/jsPsych/tree/collab_filter_dev) I spotted the error. The inability to find the plugin is a secondary error, which is happening because the plugin file contains an error. The source of the error is the first error listed in the console: "video_element is not defined", line 56 of the plugin file.
![collab-filtering-error](https://user-images.githubusercontent.com/9041788/121261902-2ce08e00-c868-11eb-8bc3-a88ae3abff5f.PNG)
You're getting this error because you're using the `video_element` JS variable in the `plugin.info` section, but `plugin.info` doesn't have access to the variables that are created in the `plugin.trial` function. To fix this, I would just change the `max_stop` parameter so that it is an integer that is separate from the `video_element` duration, and re-define the parameter to be the maximum time to stop the video before the end of the video, e.g.
```js
 max_stop: { 
  type: jsPsych.plugins.parameterType.FLOAT,
  pretty_name: 'Max stop (seconds)',
  default: 10,
  description: 'Maximum duration before the end of the video to stop the clip, in seconds. Default is ten.'
}
```
Then in `plugin.trial`, you can just subtract the `max_stop` value from the video's duration, which I think is what you're doing here anyway:
```js
// get video duration
var duration = video.duration;
//specify the max time
var max = duration - trial.videos.max_stop
```
I haven't looked at the rest of your code yet, but I'll get back to you with any suggestions for improving it when I get the chance 😃

**Comment by jackransomlovell** [comment]
Hey, @becky-gilbert thank you so much for the quick response! Yes, you're correct this is hopefully going to turn into a new plugin. Sorry if opening is an issue wasn't the right move. Feel free to change things as need be! 

Your comment is exactly right, I was initially thinking the max duration should be the duration of the stimulus but the experimenters can just change the default to 0 if so. I will change that tonight and see if it works. 

Your help is much appreciated, no rush on feedback whenever is best for you!

**Comment by jackransomlovell** [comment]
Ok so I made some changes and am now getting the following error:
```
jspsych-collab-filtering.js:234 Uncaught TypeError: Cannot read property '0' of undefined
    at Object.jsPsych.plugins.collab-filtering.plugin.trial (jspsych-collab-filtering.js:234)
    at doTrial (jspsych.js:1019)
    at nextTrial (jspsych.js:967)
    at Object.window.jsPsych.core.finishTrial (jspsych.js:339)
    at end_trial (jspsych-preload.js:339)
    at on_success (jspsych-preload.js:275)
    at file_loading_success (jspsych-preload.js:254)
    at XMLHttpRequest.request.onload (jspsych.js:2753)
```
The code at 234 is: 
```
var video_preload_blob = jsPsych.pluginAPI.getVideoBuffer(trial.videos.stimulus[0]);
```
Which comes directly from the ```jspsych-video-button-response.js``` plugin. I tried finding some documentation on it but only found something for ```.getAudioBuffer()``` which I wasn't able to make sense of. I'll keep looking and report back if I make sense of things!

**Comment by jackransomlovell** [comment]
The solution to above was indexing any instance of ```trial.video.stimulus``` as ```trial.video[0].stimulus```. I think this has to do with the nesting structure I imposed. Any how, I am now getting this following error: 

```
jspsych-collab-filtering.js:432 Uncaught TypeError: Failed to set the 'currentTime' property on 'HTMLMediaElement': The provided double value is non-finite.
```

I have a feeling this has to do with not preloading the video properly, I will make sure to report back any sort of advancements!

**Comment by becky-gilbert** [comment]
Yes I'm guessing you're right. Perhaps you've tried this already, but I would start by using `console.log` to look at the value of `pauseTime` on line 432, as well as the values that are used to generate the `pauseTime` value inside the `rand_stop` function (e.g. the video duration).
